x/ccv/provider/keeper/double_vote_test.go

```diff
@@ -0,0 +1,308 @@
+package keeper_test
+
+import (
+	"testing"
+	"time"
+
+	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	testutil "github.com/cosmos/interchain-security/v2/testutil/crypto"
+	testkeeper "github.com/cosmos/interchain-security/v2/testutil/keeper"
+	"github.com/stretchr/testify/require"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+func TestVerifyDoubleVotingEvidence(t *testing.T) {
+	keeper, ctx, ctrl, _ := testkeeper.GetProviderKeeperAndCtx(t, testkeeper.NewInMemKeeperParams(t))
+	defer ctrl.Finish()
+
+	chainID := "consumer"
+
+	signer1 := tmtypes.NewMockPV()
+	signer2 := tmtypes.NewMockPV()
+
+	val1 := tmtypes.NewValidator(signer1.PrivKey.PubKey(), 1)
+	val2 := tmtypes.NewValidator(signer2.PrivKey.PubKey(), 1)
+
+	valSet := tmtypes.NewValidatorSet([]*tmtypes.Validator{val1, val2})
+
+	blockID1 := testutil.MakeBlockID([]byte("blockhash"), 1000, []byte("partshash"))
+	blockID2 := testutil.MakeBlockID([]byte("blockhash2"), 1000, []byte("partshash"))
+
+	ctx = ctx.WithBlockTime(time.Now())
+
+	valPubkey1, err := cryptocodec.FromTmPubKeyInterface(val1.PubKey)
+	require.NoError(t, err)
+
+	valPubkey2, err := cryptocodec.FromTmPubKeyInterface(val2.PubKey)
+	require.NoError(t, err)
+
+	testCases := []struct {
+		name    string
+		votes   []*tmtypes.Vote
+		chainID string
+		pubkey  cryptotypes.PubKey
+		expPass bool
+	}{
+		{
+			"verifying public key doesn't correspond to validator address",
```

Comment:
If we were to remove the check you introduced in the verification that the address stems from the public key, this test would also have `expPass = false` because `validatorA.Address() != validatorB.Address()`. In other words, if someone accidentally removes the check you introduced, the test wouldn't complain. It might make sense to guarantee here that the different votes have the same validator address to avoid such a scenario.

Reply:
test updated [79c0492db](https://github.com/cosmos/interchain-security/pull/1360/commits/79c0492db08a60d239760a7b393d1d0ed25f43cf)